dlib/media/ffmpeg_utils.h

```diff
@@ -680,26 +665,90 @@ namespace dlib
             return name ? std::string(name) : std::string("unknown");
         }
 
+// ---------------------------------------------------------------------------------------------------
+
+        namespace details
+        {
+            inline dlib::logger& logger_internal_private()
+            {
+                static dlib::logger GLOBAL("ffmpeg.internal");
+                return GLOBAL;
+            }
+        }
+
+        inline dlib::logger& logger_internal()
+        {
+            auto& logger    = details::logger_internal_private();
+            std::ignore     = details::register_ffmpeg();
+            return logger;
+        }
+
+        inline dlib::logger& logger_dlib()
+        {
+            static dlib::logger GLOBAL("ffmpeg.dlib");
+            return GLOBAL;
+        }
+
 // ---------------------------------------------------------------------------------------------------
 
         namespace details
         {
 
 // ---------------------------------------------------------------------------------------------------
 
-            inline bool register_ffmpeg::register_library()
+            inline bool register_ffmpeg()
             {
-                avdevice_register_all();
+                static const bool REGISTERED = []
+                {
+                    avdevice_register_all();
 #if LIBAVCODEC_VERSION_INT < AV_VERSION_INT(58, 10, 100)
-                // See https://github.com/FFmpeg/FFmpeg/blob/70d25268c21cbee5f08304da95be1f647c630c15/doc/APIchanges#L91
-                avcodec_register_all();
+                    // See https://github.com/FFmpeg/FFmpeg/blob/70d25268c21cbee5f08304da95be1f647c630c15/doc/APIchanges#L91
+                    avcodec_register_all();
 #endif
 #if LIBAVFORMAT_VERSION_INT < AV_VERSION_INT(58, 9, 100) 
-                // See https://github.com/FFmpeg/FFmpeg/blob/70d25268c21cbee5f08304da95be1f647c630c15/doc/APIchanges#L86
-                av_register_all();
+                    // See https://github.com/FFmpeg/FFmpeg/blob/70d25268c21cbee5f08304da95be1f647c630c15/doc/APIchanges#L86
+                    av_register_all();
 #endif
-                return true;
-            }
+
+                    av_log_set_callback([](void* ptr, int level, const char *fmt, va_list vl) 
+                    {
+                        auto& logger = details::logger_internal_private();
+
+                        char line[256] = {0};
+                        static int print_prefix = 1;
+
+                        // Not sure if copying to vl2 is required by internal ffmpeg functions do this...
+                        va_list vl2;
+                        va_copy(vl2, vl);
+                        int size = av_log_format_line2(ptr, level, fmt, vl2, &line[0], sizeof(line), &print_prefix);
+                        va_end(vl2);
+
+                        // Remove all '\n' since dlib's logger already adds one
+                        size = std::min<int>(size, sizeof(line));
```

Comment:
I think looking at the docs for `av_log_format_line2` we should instead do:
```
size = std::min<int>(size, sizeof(line)-1);
```
then
```
line[size] = '\0'
```